browser/ntp_background/ntp_p3a_helper_impl.cc

```diff
@@ -0,0 +1,227 @@
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
+#include "base/time/time.h"
+#include "brave/components/p3a/brave_p3a_log_store.h"
+#include "brave/components/p3a/brave_p3a_service.h"
+#include "brave/components/p3a_utils/bucket.h"
+#include "components/prefs/pref_registry_simple.h"
+#include "components/prefs/pref_service.h"
+#include "components/prefs/scoped_user_pref_update.h"
+
+namespace ntp_background_images {
+
+namespace {
+
+constexpr char kNewTabPageEventCountDictPref[] =
+    "brave.brave_ads.p3a.ntp_event_count";
+
+constexpr int kCountBuckets[] = {0, 1, 2, 3, 8, 12, 16};
+
+constexpr char kViewEventKey[] = "views";
+constexpr char kClickEventKey[] = "clicks";
+constexpr char kLandEventKey[] = "lands";
+
+constexpr char kInflightDictKey[] = "inflight";
+constexpr char kExpireTimeKey[] = "expiry_time";
+
+constexpr base::TimeDelta kCountExpiryTime = base::Days(30);
+
+constexpr base::TimeDelta kStartLandingCheckTime = base::Milliseconds(750);
+constexpr base::TimeDelta kLandingTime = base::Seconds(10);
+
+}  // namespace
+
+NTPP3AHelperImpl::NTPP3AHelperImpl(PrefService* local_state,
+                                   brave::BraveP3AService* p3a_service)
+    : local_state_(local_state), p3a_service_(p3a_service) {
+  DCHECK(local_state);
+  DCHECK(p3a_service);
+  metric_sent_subscription_ =
+      p3a_service->RegisterMetricSentCallback(base::BindRepeating(
+          &NTPP3AHelperImpl::OnP3AMetricSent, base::Unretained(this)));
+  rotation_subscription_ =
+      p3a_service->RegisterRotationCallback(base::BindRepeating(
+          &NTPP3AHelperImpl::OnP3ARotation, base::Unretained(this)));
+}
+
+NTPP3AHelperImpl::~NTPP3AHelperImpl() = default;
+
+void NTPP3AHelperImpl::RegisterLocalStatePrefs(PrefRegistrySimple* registry) {
+  registry->RegisterDictionaryPref(kNewTabPageEventCountDictPref);
+}
+
+void NTPP3AHelperImpl::RecordView(const std::string& creative_instance_id) {
+  if (!p3a_service_->IsP3AEnabled()) {
+    return;
+  }
+  UpdateMetricCount(creative_instance_id, kViewEventKey);
+}
+
+void NTPP3AHelperImpl::RecordClickAndMaybeLand(
+    const std::string& creative_instance_id) {
+  if (!p3a_service_->IsP3AEnabled()) {
+    return;
+  }
+  UpdateMetricCount(creative_instance_id, kClickEventKey);
+  landing_check_timer_.Start(
+      FROM_HERE, kStartLandingCheckTime,
+      base::BindOnce(&NTPP3AHelperImpl::OnLandingStartCheck,
+                     base::Unretained(this), creative_instance_id));
+}
+
+void NTPP3AHelperImpl::SetLastTabURL(const GURL& url) {
+  last_tab_hostname_ = url.host();
+}
+
+void NTPP3AHelperImpl::OnP3ARotation(bool is_express) {
+  if (!is_express) {
+    return;
+  }
+  DictionaryPrefUpdate update(local_state_, kNewTabPageEventCountDictPref);
+  base::Value::Dict& update_dict = update->GetDict();
+
+  if (!p3a_service_->IsP3AEnabled()) {
+    update_dict.clear();
+    return;
+  }
+
+  for (const auto [creative_instance_id, creative_value] : update_dict) {
```

Comment:
const auto& []

Reply:
I don't think that's possible when binding to a temporary value (at least that's what the compiler tells me).

Reply:
confirmed over dm